### PR TITLE
MVI implementation example

### PR DESCRIPTION
**Pull request by MDikkii**

### CODE REVIEW

File: app/src/main/java/com/example/appName/presentation/features/main/MainPresenter.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+                previousState.copy(mainText = partialState.loggedText, isLoggedIn = true)
+        }
+    }
+
+    override fun provideViewIntents(): List<Observable<MainViewState.PartialState>> = listOf(
+            onLoginIntent(),
+            onLogoutIntent()
+    )
+
+    private fun onLogoutIntent(): Observable<MainViewState.PartialState> =
```
Comment (kubak89) [inline app/src/main/java/com/example/appName/presentation/features/main/MainPresenter.kt]: I think that function names should reflect their purpose rather then inform _when_ they're called, so better name for that method would be `logout` in my opinion.

File: app/src/main/java/com/example/appName/presentation/features/main/MainPresenter.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+            view.logoutIntent.map { MainViewState.PartialState.WelcomeState(welcomeText = getGreetingsText(WELCOME_NAME)) }
+
+    private fun onLoginIntent(): Observable<MainViewState.PartialState> =
```
Comment (kubak89) [inline app/src/main/java/com/example/appName/presentation/features/main/MainPresenter.kt]: Same as above.

File: app/src/main/java/com/example/appName/presentation/features/main/MainPresenter.kt [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            view.loginIntent.switchMap {
+                exampleUserRepository.getUser().toObservable()
+            }.map { user ->
+                MainViewState.PartialState.LoggedState(loggedText = getGreetingsText(user.name))
+            }
+
+    private fun getGreetingsText(name: String) = "Welcome $name!"
+
+    companion object {
+        const val WELCOME_NAME = "Stranger"
```
Comment (kubak89) [inline app/src/main/java/com/example/appName/presentation/features/main/MainPresenter.kt]: Nit, but I think `DEFAULT_WELCOME_NAME` or `LOGGED_OUT_NAME` would be better.

File: app/src/main/java/com/example/appName/presentation/features/main/MainViewState.kt [hunk 4]
```diff
@@ -0,0 +1,13 @@
+package com.example.appName.presentation.features.main
+
+import java.io.Serializable
+
+data class MainViewState(
+        val mainText: String,
+        val isLoggedIn: Boolean
+) : Serializable {
+    sealed class PartialState {
+        class WelcomeState(val welcomeText: String) : PartialState()
```
Comment (kubak89) [inline app/src/main/java/com/example/appName/presentation/features/main/MainViewState.kt]: Data class would be better here for easier debugging (it generates `toString` method).

File: app/src/main/java/com/example/appName/presentation/features/main/MainModule.kt [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+@Module
+class MainModule {
+
+    @Provides
+    fun provideMainView(activity: MainActivity): MainView = activity
+
+    @Provides
+    fun provideInitialMainViewState(activity: MainActivity): MainViewState = activity.savedInstanceState?.getSerializable(
+            BaseActivity.KEY_SAVED_ACTIVITY_VIEW_STATE) as? MainViewState
+            ?: MainViewState(mainText = "Welcome Stranger!", isLoggedIn = false)
```
[the remark this answers is not in the document]
Comment (MDikkii) [reply app/src/main/java/com/example/appName/presentation/features/main/MainModule.kt]: Done

File: app/src/main/java/com/example/appName/presentation/features/main/MainPresenter.kt [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import javax.inject.Inject
+
+class MainPresenter @Inject constructor(
+        private val view: MainView,
+        initialState: MainViewState,
+        private val exampleUserRepository: ExampleUserRepository
+) : BasePresenter<MainViewState, MainViewState.PartialState>(initialState) {
+
+    override fun reduceViewState(previousState: MainViewState, partialState: MainViewState.PartialState): MainViewState {
+        return when (partialState) {
```
[the remark this answers is not in the document]
Comment (MDikkii) [reply app/src/main/java/com/example/appName/presentation/features/main/MainPresenter.kt]: Right

File: app/src/main/java/com/example/appName/presentation/features/main/MainPresenter.kt (hunk 1 above)
Comment (MDikkii) [reply app/src/main/java/com/example/appName/presentation/features/main/MainPresenter.kt]: Done, thanks

File: app/src/main/java/com/example/appName/presentation/features/main/MainPresenter.kt (hunk 2 above)
Comment (MDikkii) [reply app/src/main/java/com/example/appName/presentation/features/main/MainPresenter.kt]: Done, thanks

File: app/src/main/java/com/example/appName/presentation/features/main/MainActivity.kt [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+    //region Render methods
+    override fun render(viewState: MainViewState) {
+        with(viewState) {
+            updateText(mainText)
+            updateButtonsVisibility(isLoggedIn)
+        }
+    }
+
+    private fun updateText(text: String) {
```
[the remark this answers is not in the document]
Comment (MDikkii) [reply app/src/main/java/com/example/appName/presentation/features/main/MainActivity.kt]: Renamed

File: app/src/main/java/com/example/appName/presentation/features/main/MainActivity.kt [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        main_text.text = text
+    }
+
+    private fun updateButtonsVisibility(isLoggedIn: Boolean) {
```
[the remark this answers is not in the document]
Comment (MDikkii) [reply app/src/main/java/com/example/appName/presentation/features/main/MainActivity.kt]: Renamed

File: app/src/main/java/com/example/appName/presentation/features/main/MainPresenter.kt (hunk 3 above)
Comment (MDikkii) [reply app/src/main/java/com/example/appName/presentation/features/main/MainPresenter.kt]: Done, moved to Constant object to use it also in view state

File: app/src/main/java/com/example/appName/presentation/features/main/MainViewState.kt (hunk 4 above)
Comment (MDikkii) [reply app/src/main/java/com/example/appName/presentation/features/main/MainViewState.kt]: Done

File: app/src/main/java/com/example/appName/presentation/features/main/MainActivity.kt [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        bindIntents()
+    }
+
+    private fun bindIntents() {
+        login_button.setOnClickListener {
+            loginIntent.onNext(Unit)
+        }
+
+        logout_button.setOnClickListener {
+            logoutIntent.onNext(Unit)
```
[the remark this answers is not in the document]
Comment (MDikkii) [reply app/src/main/java/com/example/appName/presentation/features/main/MainActivity.kt]: done

File: app/src/main/java/com/example/appName/presentation/features/main/MainActivity.kt [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    }
+    //endregion
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        bindIntents()
+    }
+
+    private fun bindIntents() {
+        login_button.setOnClickListener {
```
[the remark this answers is not in the document]
Comment (MDikkii) [reply app/src/main/java/com/example/appName/presentation/features/main/MainActivity.kt]: done

File: app/src/main/java/com/example/appName/presentation/features/main/MainActivity.kt [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+import com.jakewharton.rxbinding3.view.clicks
+import io.reactivex.Observable
+import kotlinx.android.synthetic.main.activity_main.*
+
+class MainActivity : BaseActivity<MainViewState, MainPresenter>(
+        R.layout.activity_main
+), MainView {
+    //region Intents
+    override val loginIntent: Observable<Unit>
+        get() = login_button.clicks()
```
Comment (przemek19980102) [inline app/src/main/java/com/example/appName/presentation/features/main/MainActivity.kt]: shouldn't this be `by lazy`? Don't you have issues with login button not being initialized?

File: app/src/main/java/com/example/appName/presentation/features/main/MainActivity.kt [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    override val logoutIntent: Observable<Unit>
+        get() = logout_button.clicks()
+    //endregion
+
+    //region Render methods
+    override fun render(viewState: MainViewState) {
+        with(viewState) {
+            renderText(mainText)
```
Comment (przemek19980102) [inline app/src/main/java/com/example/appName/presentation/features/main/MainActivity.kt]: I'm a fan of passing a full viewState to a specific render function, but this is okay in such example

File: app/src/main/java/com/example/appName/presentation/features/main/MainPresenter.kt [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+        private val exampleUserRepository: ExampleUserRepository
+) : BasePresenter<MainViewState, MainViewState.PartialState>(initialState) {
+
+    override fun reduceViewState(previousState: MainViewState, partialState: MainViewState.PartialState): MainViewState =
+            when (partialState) {
+                is MainViewState.PartialState.WelcomeState ->
+                    previousState.copy(mainText = partialState.welcomeText, isLoggedIn = false)
+                is MainViewState.PartialState.LoggedInState ->
+                    previousState.copy(mainText = partialState.loggedInText, isLoggedIn = true)
+
```
Comment (przemek19980102) [inline app/src/main/java/com/example/appName/presentation/features/main/MainPresenter.kt]: blank line here

File: app/src/main/java/com/example/appName/presentation/features/main/MainPresenter.kt [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+            }
+
+    override fun provideViewIntents(): List<Observable<MainViewState.PartialState>> = listOf(
+            login(),
```
Comment (przemek19980102) [inline app/src/main/java/com/example/appName/presentation/features/main/MainPresenter.kt]: I'd suggest createLoginIntent() and createLogoutIntent()

File: app/src/main/java/com/example/appName/presentation/features/main/MainPresenter.kt [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+            view.logoutIntent.map { MainViewState.PartialState.WelcomeState(welcomeText = getGreetingsText(LOGGED_OUT_NAME)) }
+
+    private fun login(): Observable<MainViewState.PartialState> =
+            view.loginIntent.switchMap {
+                exampleUserRepository.getUser().toObservable()
+            }.map { user ->
+                MainViewState.PartialState.LoggedInState(loggedInText = getGreetingsText(user.name))
+            }
+
+    private fun getGreetingsText(name: String) = "Welcome $name!"
```
Comment (przemek19980102) [inline app/src/main/java/com/example/appName/presentation/features/main/MainPresenter.kt]: shouldn't this be handled on UI side instead?

Comment (MDikkii) [reply app/src/main/java/com/example/appName/presentation/features/main/MainPresenter.kt]: Right, moved to render method

File: app/src/main/java/com/example/appName/presentation/features/main/MainPresenter.kt (hunk 13 above)
Comment (MDikkii) [reply app/src/main/java/com/example/appName/presentation/features/main/MainPresenter.kt]: Eagle eye 🥇 done

File: app/src/main/java/com/example/appName/presentation/features/main/MainActivity.kt (hunk 12 above)
Comment (MDikkii) [reply app/src/main/java/com/example/appName/presentation/features/main/MainActivity.kt]: I haven't strong feelings about it, maybe passing the whole viewState make it easier to render something more complex, changed as proposed.

**Pull request merged**